### PR TITLE
Implement provider

### PR DESCRIPTION
**Pull request by AndreyBelym**

### CODE REVIEW

**Comment by AndreyBelym** [comment]
It's ready for review, but needs `testcafe` with `testcafe-hammerhead@10.3.4`.
\cc @AlexanderMoskovkin @kirovboris @churkin

**Comment by AlexanderMoskovkin** [comment]
`testcafe@0.13.0` uses `testcafe-hammerhead@10.3.4`

File: src/utils/get-config.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { readFileSync } from 'fs';
+import path from 'path';
+import resolveFileUrl from './resolve-file-url';
+
+
+export default function (appPath) {
+    var configPath   = path.join(appPath, '.testcafe-electron-rc');
+    var configString = readFileSync(configPath).toString();
```
Comment (AlexanderMoskovkin) [inline src/utils/get-config.js]: I think we should throw an error with message like `Place .testcafe-electron-rc file to the root of your application`  if the config files doesn't exist

File: README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+This is the **electron** browser provider plugin for [TestCafe](http://devexpress.github.io/testcafe).
+
+## Install
+
+```
+npm install testcafe-browser-provider-electron
+```
+
+## Usage
+First, create a `.testcafe-electron.json` file in root directory of your Electron app:
```
Comment (AlexanderMoskovkin) [inline README.md]: It seems it should be `.testcafe-electron-rc` (https://github.com/AndreyBelym/testcafe-browser-provider-electron/pull/1/files#diff-ae8cda07c4f25ff93a279e308d74a7f0R7)

**Comment by AndreyBelym** [comment]
\cc @AlexanderMoskovkin @kirovboris @inikulin

File: README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    .createRunner()
+    .src('path/to/test/file.js')
+    .browsers('electron:/home/user/electron-app')
+    .run();
+```
+
+## Config
+Below is complete list of all configuration options that you can put into `.testcafe-electron-rc`.
+
+ - `mainWindowUrl` __(mandatory)__ - specifies URL used for the main window page of the appplication.
```
Comment (inikulin) [inline README.md]: `(mandatory)` -> `(required)`

File: README.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ - `disableNavigateEvents` __(optional)__ - if you use `did-navigate` ow `will-navigate` webContent events to prevent navigation, you should disable it by setting this option to `true`.
+ - `openDevTools` __(optional)__ - if `true`, DevTools will be opened just before tests start.
+ 
+## Helpers
+You can use some helper functions from provider in your test files. Use ES6 import statement to get them, like 
+```js
+import { mainMenu, menuClick } from 'testcafe-browser-provider-electron';
+```
+
+ - `async function mainMenu ()` - get a snapshot of application main menu. You can check properties available in the snapshot [here](https://github.com/electron/electron/blob/master/docs/api/menu-item.md),
```
Comment (inikulin) [inline README.md]: It's better be named `getMainMenu` and `getContextMenu`

File: README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ 
+ - `async function contextMenu ()` - get a snapshot of context menu. You can check properties available in the snapshot [here](https://github.com/electron/electron/blob/master/docs/api/menu-item.md), 
+
+ - `async function menuClick (menuSnapshot, modifiers)` - perform a click on the given `menuSnapshot`. You can pass state of control keys (`Ctrl`, `Alt`, `Meta` etc.) in the `modifiers` argument.
```
Comment (inikulin) [inline README.md]: Clarify how to click buttons in menu

**Comment by AndreyBelym** [comment]
FPR

**Comment by AndreyBelym** [comment]
\сс @VasilyStrelyaev

File: README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ [here](https://github.com/electron/electron/blob/master/docs/api/menu.md). 
+ 
+ - `async function getContextMenu ()` - get a snapshot of context menu. You can check properties available in the snapshot 
+ [here](https://github.com/electron/electron/blob/master/docs/api/menu.md), 
+
+ - `async function clickOnMenuItem (menuItem, modifiers)` - perform a click on the given `menuItem`. It can be a string, 
+ in this case it will be passed to the `getMenuItem` function and the returned value will be used; or a value retrieved 
+ with `getMenuItem`, `getMainMenu`, `getContextMenu` functions. 
+ Also you can pass state of control keys (`Ctrl`, `Alt`, `Meta` etc.) in the `modifiers` argument, e.g. the default is 
+ `{ shiftKey: false, ctrlKey: false, altKey: false, metaKey: false}`. Examples: `clickOnMenuItem('Main Menu > File > Open')`,
```
Comment (AlexanderMoskovkin) [inline README.md]: let's rename modifiers:
```js
modifiers: { ctrlKey, shiftKey, ... }   ---> modifiers: { ctrl, shift, ... }
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+This is the **electron** browser provider plugin for [TestCafe](http://devexpress.github.io/testcafe).
+
+## Install
+
+```
+npm install testcafe-browser-provider-electron
+```
+
+## Usage
+First, create a `.testcafe-electron-rc` file in root directory of your Electron app (for more info see the Config section):
```
Comment (VasilyStrelyaev) [inline README.md]:
```
First, create a `.testcafe-electron-rc` file in the root directory of your Electron app. For more info, see the Config section.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+```
+{
+  "mainWindowUrl": "./index.html"
+}
+```
+
+When you run tests from the command line, specify path to Electron app prefixed with "electron:" -
```
Comment (VasilyStrelyaev) [inline README.md]:
```
When you run tests from the command line, specify the path to the Electron app prefixed with "electron:" -
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+```js
+testCafe
+    .createRunner()
+    .src('path/to/test/file.js')
+    .browsers('electron:/home/user/electron-app')
+    .run();
+```
+
+## Config
+Below is complete list of all configuration options that you can put into `.testcafe-electron-rc`.
```
Comment (VasilyStrelyaev) [inline README.md]:
```
Below is a list of configuration options that you can specify in `.testcafe-electron-rc`.
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+ - `mainWindowUrl` __(required)__ - specifies URL used for the main window page of the appplication. 
+ If you use `file://` urls, you can also specify a relative (to the application directory) or an absolute path to the file of the page.
```
Comment (VasilyStrelyaev) [inline README.md]:
```
 - `mainWindowUrl` __(required)__ - specifies the URL used for the application's main window page. 
 If you use `file://` urls, you can also specify a relative (to the application directory) or an absolute path to the page file.
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+ - `appPath` __(optional)__ - alters path to the application. By default, application path is the part after `electron:` 
+ of the string used in TestCafe CLI or API. You can override it by specifying an absolute path, or append a relative path from 'appPath'.
```
Comment (VasilyStrelyaev) [inline README.md]:
```
 - `appPath` __(optional)__ - alters path to the application. By default, the application path is specified after the `electron:` prefix in the string passed to the command line or API. You can override it by specifying an absolute path, or append a relative path from 'appPath'. 
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+## Config
+Below is complete list of all configuration options that you can put into `.testcafe-electron-rc`.
+
+ - `mainWindowUrl` __(required)__ - specifies URL used for the main window page of the appplication. 
+ If you use `file://` urls, you can also specify a relative (to the application directory) or an absolute path to the file of the page.
+ - `appPath` __(optional)__ - alters path to the application. By default, application path is the part after `electron:` 
+ of the string used in TestCafe CLI or API. You can override it by specifying an absolute path, or append a relative path from 'appPath'. 
+ - `appArgs` __(optional)__ - overrides application commandline arguments with the values specified in this option. It should be an array or an object with numeric keys.
```
Comment (VasilyStrelyaev) [inline README.md]: command line

File: README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+ - `disableNavigateEvents` __(optional)__ - if you use `did-navigate` ow `will-navigate` webContent events to prevent navigation, you should disable it by setting this option to `true`.
```
Comment (VasilyStrelyaev) [inline README.md]: > ow

or

File: README.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+ - `openDevTools` __(optional)__ - if `true`, DevTools will be opened just before tests start.
+ 
+## Helpers
+You can use some helper functions from provider in your test files. Use ES6 import statement to get them, like
```
Comment (VasilyStrelyaev) [inline README.md]:
```
You can use helper functions from the provider in your test files. Use ES6 import statement to access them.
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+## Helpers
+You can use some helper functions from provider in your test files. Use ES6 import statement to get them, like 
+```js
+import { getMainMenu, clickOnMenuItem } from 'testcafe-browser-provider-electron';
+```
+ - `async function getMenuItem (menuItemSelector)` - get a snapshot of the given menu item. `menuItemSelector` is a string that consists
+ of menu type and menu item labels, separated by the `>` sign, e.g. `Main Menu > File > Open` or `Context Menu > Undo`. 
+ The `Main Menu` menu type can be skipped. If there are a number of the specified menu items with the same label on the same level, 
+ you can specify a one-based index in square brackets, e.g. `Main Menu > Window > My Window [2]` selects the second menu item with 
+ label `My Window` in the `Window` menu. Check properties available in the snapshot [here](https://github.com/electron/electron/blob/master/docs/api/menu-item.md).
```
Comment (VasilyStrelyaev) [inline README.md]:
```
 - `async function getMenuItem (menuItemSelector)` - gets a snapshot of the specified menu item. `menuItemSelector` is a string that consists
 of the menu type and menu item labels, separated by the `>` sign, e.g. `Main Menu > File > Open` or `Context Menu > Undo`. 
 The `Main Menu` menu type can be skipped. If there are several menu items with the same label on the same level,     
 you can specify a one-based index in square brackets, e.g. `Main Menu > Window > My Window [2]` selects the second menu item with 
 label `My Window` in the `Window` menu. Check the properties available in the snapshot [here](https://github.com/electron/electron/blob/master/docs/api/menu-item.md).
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+ - `async function getMainMenu ()` - get a snapshot of application main menu. You can check properties available in the snapshot
```
Comment (VasilyStrelyaev) [inline README.md]: >get a snapshot of application main menu

gets a snapshot of the application main menu

File: README.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+ - `async function getMenuItem (menuItemSelector)` - get a snapshot of the given menu item. `menuItemSelector` is a string that consists
+ of menu type and menu item labels, separated by the `>` sign, e.g. `Main Menu > File > Open` or `Context Menu > Undo`. 
+ The `Main Menu` menu type can be skipped. If there are a number of the specified menu items with the same label on the same level, 
+ you can specify a one-based index in square brackets, e.g. `Main Menu > Window > My Window [2]` selects the second menu item with 
+ label `My Window` in the `Window` menu. Check properties available in the snapshot [here](https://github.com/electron/electron/blob/master/docs/api/menu-item.md).
+
+ - `async function getMainMenu ()` - get a snapshot of application main menu. You can check properties available in the snapshot 
+ [here](https://github.com/electron/electron/blob/master/docs/api/menu.md). 
+ 
+ - `async function getContextMenu ()` - get a snapshot of context menu. You can check properties available in the snapshot
```
Comment (VasilyStrelyaev) [inline README.md]: > get a snapshot of context menu

gets a snapshot of the context menu

File: README.md [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ - `async function getMainMenu ()` - get a snapshot of application main menu. You can check properties available in the snapshot 
+ [here](https://github.com/electron/electron/blob/master/docs/api/menu.md). 
+ 
+ - `async function getContextMenu ()` - get a snapshot of context menu. You can check properties available in the snapshot 
+ [here](https://github.com/electron/electron/blob/master/docs/api/menu.md), 
+
+ - `async function clickOnMenuItem (menuItem, modifiers)` - perform a click on the given `menuItem`. It can be a string, 
+ in this case it will be passed to the `getMenuItem` function and the returned value will be used; or a value retrieved 
+ with `getMenuItem`, `getMainMenu`, `getContextMenu` functions. 
+ Also you can pass state of control keys (`Ctrl`, `Alt`, `Meta` etc.) in the `modifiers` argument, e.g. the default is
```
Comment (VasilyStrelyaev) [inline README.md]: > perform 

performs 

> Also you can pass state of control keys

Also, you can pass the state of the control keys

File: README.md [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+ - `async function clickOnMenuItem (menuItem, modifiers)` - perform a click on the given `menuItem`. It can be a string, 
+ in this case it will be passed to the `getMenuItem` function and the returned value will be used; or a value retrieved 
+ with `getMenuItem`, `getMainMenu`, `getContextMenu` functions. 
+ Also you can pass state of control keys (`Ctrl`, `Alt`, `Meta` etc.) in the `modifiers` argument, e.g. the default is 
+ `{ shift: false, ctrl: false, alt: false, meta: false}`. Examples: `clickOnMenuItem('Main Menu > File > Open')`,
+ `clickOnMenuItem('File > Open')`, `clickOnMenuItem((await getMainMenu()).items[0].submenu.items[0])`,
+ 
+ - `async function setElectronDialogHandler (handler, dependencies)` - set a function `function handler (type, ...args)` that will handle native Electron dialogs. Specify global variables of the function in the `dependencies` argument. 
+ Handler function must be synchronous and will be invoked with the dialog type `type`, and the arguments `args` from the original dialog function.
```
Comment (VasilyStrelyaev) [inline README.md]:
```
 - `async function setElectronDialogHandler (handler, dependencies)` - sets a function `function handler (type, ...args)` that will handle native Electron dialogs. Specify the function's global variables in the `dependencies` argument. 
 The handler function must be synchronous and will be invoked with the dialog type `type`, and the arguments `args` from the original dialog function.
```

**Comment by AndreyBelym** [comment]
FPR

**Pull request merged**